components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt

```diff
@@ -0,0 +1,338 @@
+        try {
+            val localDevice = block()
+            sendEvent(FxaEvent.DeviceOperationComplete(operation, localDevice))
+            result?.complete(true)
+        } catch(e: FxaException) {
+            Log.e(LOG_TAG, "Exception when handling $action", e)
+            sendEvent(FxaEvent.DeviceOperationFailed(operation))
+            result?.complete(false)
+        }
+    }
+
+    private suspend fun handleDisconnect(action: FxaAction.Disconnect) {
+        if (action.fromAuthIssues) {
+            inner.disconnectFromAuthIssues()
+            tryPersistState()
+            changeState(FxaAuthState.Disconnected(fromAuthIssues=true), FxaAuthStateTransition.AUTH_CHECK_FAILED)
+        } else {
+            inner.disconnect()
+            tryPersistState()
+            changeState(FxaAuthState.Disconnected(), FxaAuthStateTransition.DISCONNECTED)
+        }
+    }
+
+    private suspend fun handleCheckAuthorization(currentState: FxaAuthState.Connected) {
+        changeState(currentState.copy(authCheckInProgress=true), FxaAuthStateTransition.AUTH_CHECK_STARTED)
+        val success = try {
+            val status = withNetworkRetry { inner.checkAuthorizationStatus() }
+            status.active
+        } catch(e: FxaException.Authentication) {
+            // The Rust code should handle this exception, but if it doesn't, let's consider it a
+            // failed check.
+            Log.e(LOG_TAG, "Authentication exception when checking authorization status", e)
+            false
+        } catch(e: FxaException) {
+            Log.e(LOG_TAG, "Exception when checking authorization status", e)
+            // It's not clear what's the Right Thing to do in this case, but considering it a success is
+            // better than a failure.  We don't want users logged out because of network issues.
+            true
+        }
+        if (success) {
+            tryPersistState()
+            changeState(currentState.copy(authCheckInProgress=false), FxaAuthStateTransition.AUTH_CHECK_SUCCESS)
+        } else {
+            inner.disconnectFromAuthIssues()
+            tryPersistState()
+            changeState(FxaAuthState.Disconnected(true), FxaAuthStateTransition.AUTH_CHECK_FAILED)
+        }
+    }
+}
+
+internal class RetryLogic {
+    private var lastNetworkRetry: Long = 0
+    private var lastAuthCheck: Long = 0
+
+    fun shouldRetryAfterNetworkError(): Boolean {
+        val elasped = (System.currentTimeMillis() - lastNetworkRetry).milliseconds
+        if (elasped > 30.seconds) {
+            lastNetworkRetry = System.currentTimeMillis()
+            return true
+        } else {
+            return false
+        }
+    }
+
+    fun shouldRecheckAuthStatus(): Boolean {
+        val elasped = (System.currentTimeMillis() - lastAuthCheck).milliseconds
+        if (elasped > 60.seconds) {
+            lastAuthCheck = System.currentTimeMillis()
+            return true
+        } else {
+            return false
+        }
+    }
```

Comment:
I like this class because it isolates the retry logic, but I don't really know what the correct logic is.

The old android-components code performed 3 network retries for every operation, which seemed quite a lot to me.

Auth checking had a complicated system that boiled down to don't do it too much and definitely don't get stuck in an infinite loop.  Infinite loops shouldn't be an issue anymore since we don't queue a new action to handle the auth check.